crates/driver/src/simulator/mod.rs

```diff
@@ -0,0 +1,62 @@
+use {
+    crate::logic::{competition::solution, eth},
+    thiserror::Error,
+};
+
+mod node;
+mod tenderly;
+
+#[derive(Debug, Error)]
+#[error("simulation error")]
+pub struct Error;
+
+/// Ethereum transaction simulator.
+#[derive(Debug)]
+pub struct Simulator(Inner);
+
+impl Simulator {
+    /// Simulate an onchain settlement on the Ethereum network.
+    pub async fn simulate(
+        &self,
+        _settlement: &solution::Settlement,
+    ) -> Result<eth::Simulation, Error> {
+        // TODO This should do the two-step access list generation and gas estimation
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+enum Inner {
+    Tenderly {
+        tenderly: tenderly::Tenderly,
+        node: node::Node,
+    },
+    Node(node::Node),
+}
```

Comment:
I think the drawbacks you outlines regarding `dyn` and generics are definitely valid and the approach you chose to avoid both definitely seems appealing. I just have a few gripes with that:
1. IMO it would be great if we could reduce the number of concepts we use in the code base to a minimum and agree on a minimum set of them. In this case your approach looks like it could generally replace the usage of `dyn` in our code base.
2. If we would like to replace the usage of `dyn` with this pattern I'm a bit worried that this pattern would look quite nasty for cases where there are lots of types that implement a trait (e.g. `PriceEstimating`).

I looked a bit around and it seems enough people share my concern about maintenance that someone created a crate that takes care of all the boiler plate code and makes it look like the `dyn` code we have right now where we implement traits on structs in stead of the enum itself.
I reimplemented the gist of the `simulator` logic with the [`enum_dispatch`](https://docs.rs/enum_dispatch/latest/enum_dispatch/) crate and it looks reasonable enough to me that we could liberally replace the current usage of `Box<dyn MyTrait>` with the new concept.

<details>
<summary>with enum_dispatch</summary>

```rust
#![allow(dead_code)]
use async_trait::async_trait;
use enum_dispatch::enum_dispatch;

/// crates/driver/src/simulator/mod.rs
#[async_trait]
#[enum_dispatch(Simulator)]
trait Simulating {
    async fn simulate(&self, _settlement: &[u8]);
}

#[enum_dispatch(Simulator)]
#[derive(Debug, Clone)]
enum Simulator {
    Node,
    Tenderly,
}

/// crates/driver/src/simulator/node.rs
#[derive(Debug, Clone)]
struct Node {
    url: String,
}

impl Node {
    pub fn new(url: String) -> Self {
        Self { url }
    }
}

#[async_trait]
impl Simulating for Node {
    async fn simulate(&self, _settlement: &[u8]) {
        println!("simulated with node");
    }
}

/// crates/driver/src/simulator/tenderly.rs
#[derive(Debug, Clone)]
struct Tenderly {
    url: String,
    key: String,
}

impl Tenderly {
    pub fn new(url: String, key: String) -> Self {
        Self { url, key }
    }
}

#[async_trait]
impl Simulating for Tenderly {
    async fn simulate(&self, _settlement: &[u8]) {
        println!("simulated with tenderly");
    }
}

/// crates/driver/src/main.rs
#[tokio::main]
async fn main() {
    let simulators = vec![
        Simulator::Node(Node::new("node".into())),
        Simulator::Tenderly(Tenderly::new("a".into(), "b".into())),
    ];
    let settlement = vec![1, 2, 3, 4, 5, 6];
    for simulator in &simulators {
        simulator.simulate(&settlement).await;
    }
}
```
</details>